api/src/routes/challenge.ts

```diff
@@ -166,5 +171,69 @@ export const challengeRoutes: FastifyPluginCallbackTypebox = (
     }
   );
 
+  fastify.post(
+    '/modern-challenge-completed',
+    {
+      schema: schemas.modernChallengeCompleted,
+      errorHandler(error, request, reply) {
+        if (error.validation) {
+          void reply.code(400);
+          return formatValidationError(error.validation);
+        } else {
+          fastify.errorHandler(error, request, reply);
+        }
+      }
+    },
+    async (req, reply) => {
+      try {
+        const { id, files, challengeType } = req.body;
+
+        const user = await fastify.prisma.user.findUniqueOrThrow({
+          where: { id: req.session.user.id }
+        });
+        const progressTimestamps = user.progressTimestamps as
+          | ProgressTimestamp[]
+          | null;
```

Comment:
Is this the best approach?

The type issue is because the DB isn't normalised, right? So what happens if we get an array of something other than ProgressTimestamp?

Reply:
This type should include all the types in the db, so it's not a dangerous typecast, but that could be made explicit. Calling it something like `RawProgressTimestamp` might help.